Sources/SwiftDocC/Infrastructure/Link Resolution/ExternalPathHierarchyResolver.swift

```diff
@@ -0,0 +1,235 @@
+    ///   - isCurrentlyResolvingSymbolLink: Whether or not the documentation link is a symbol link.
+    /// - Returns: The result of resolving the reference.
+    func resolve(_ unresolvedReference: UnresolvedTopicReference, fromSymbolLink isCurrentlyResolvingSymbolLink: Bool) -> TopicReferenceResolutionResult {
+        do {
+            let foundID = try pathHierarchy.find(path: Self.path(for: unresolvedReference), parent: nil, onlyFindSymbols: true)
+            guard let foundReference = resolvedReference[foundID] else {
+                fatalError("Every identifier in the path hierarchy has a corresponding reference in the wrapping resolver. If it doesn't that's an indication that the file content that it was deserialized from was malformed.")
+            }
+            
+            guard content[foundReference] != nil else {
+                return .failure(unresolvedReference, .init("Resolved \(foundReference.url.withoutHostAndPortAndScheme().absoluteString.singleQuoted) but don't have any content to display for it."))
+            }
+            
+            return .success(foundReference)
+        } catch let error as PathHierarchy.Error {
+            var originalReferenceString = unresolvedReference.path
+            if let fragment = unresolvedReference.topicURL.components.fragment {
+                originalReferenceString += "#" + fragment
+            }
+            
+            return .failure(unresolvedReference, error.asTopicReferenceResolutionErrorInfo(originalReference: originalReferenceString) { node in 
+                guard let reference = resolvedReference[node.identifier], let summary = content[reference] else {
+                    return node.name
+                }
+                return summary.title
+            })
+        } catch {
+            fatalError("Only PathHierarchy.Error errors are raised from the symbol link resolution code above.")
+        }
+    }
+    
+    private static func path(for unresolved: UnresolvedTopicReference) -> String {
+        guard let fragment = unresolved.fragment else {
+            return unresolved.path
+        }
+        return "\(unresolved.path)#\(urlReadableFragment(fragment))"
+    }
+
+    /// Returns the external entity for a symbol's unique identifier or `nil` if that symbol isn't known in this external context.
+    func entity(symbolID usr: String) -> ExternalEntity? {
+        // TODO: Resolve external symbols by USR (rdar://116085974) (There is nothing calling this function)
+        // This function has an optional return value since it's not easy to check what module a symbol belongs to based on its identifier.
+        guard let reference = symbols[usr] else { return nil }
+        return entity(reference)
+    }
+    
+    /// Returns the external entity for a reference that was successfully resolved by this external resolver.
+    ///
+    /// - Important: Passing a resolved reference that wasn't resolved by this resolver will result in a fatal error.
```

Comment:
Since anything can create a `ResolvedRenderReference` there's nothing really stopping the rest of the code from passing a local reference to an external resolver. 

It may be possible that we could remove this function all together and return everything all at once from `resolve(...)` but that would need its own result type instead of `TopicReferenceResolutionResult`. 

Since both those are internal functions we could combined them in the future without breaking anyone else.